Author: Akshay Kumar

This notebook contains Questions 1, 2 & 3 of the Assignment. They have been segregated by Section headers

In [1]:

import pandas as pd

Question 1
Importing Data

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [6]:
wiki_data = wiki_data[0]
wiki_data


Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

Question 2
Installing geocoder

In [13]:
pip install geocoder

     |████████████████████████████████| 98 kB 9.3 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [14]:
import geocoder

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data.

Question 3
Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [21]:
from geopy.geocoders import Nominatim

In [22]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [24]:
!pip install folium
import folium

     |████████████████████████████████| 94 kB 6.7 MB/s  eta 0:00:01


In [25]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [29]:
CLIENT_ID = 'BLNQBYIWWQGCUFUZIV12HXWPYPA0C0D24RLWUKVM1421450G' 
CLIENT_SECRET = 'JLFQP4ICTMENWXM1MMKYD2JYINZ4U3K2LYGQUJIKVR3X2T2P'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: BLNQBYIWWQGCUFUZIV12HXWPYPA0C0D24RLWUKVM1421450G
CLIENT_SECRET:JLFQP4ICTMENWXM1MMKYD2JYINZ4U3K2LYGQUJIKVR3X2T2P


In [30]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [31]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [32]:
venues_in_toronto.shape

(1334, 5)

So we have 1317 records and 5 columns. Checking sample data

In [34]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

In [35]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1320  Mimico NW, The Queensway West, South of Bloor,...   
1321  Mimico NW, The Queensway West, South of Bloor,...   
1322  Mimico NW, The Queensway West, South of Bloor,...   
1323  Mimico NW, The Queensway West, South of Bloor,...   
1324  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1320               43.628841               -79.520999              Wingporium   
1321               43.628841               -79.520999        South St. Burger   
1322               43.628841               -79.520999               Dollarama   
1323               43.628841               -79.520999          Healthy Planet   
1324               43.628841               -79.520999                7-Eleven   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1320            Wings Joint  
1321           Burger Joint  
1322         Discount Store  
1323        Supplement Shop  
1324      Convenience Store  

[407 rows x 5 columns]

In [36]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[236 rows x 4 columns]

One Hot encoding the venue Categories

In [37]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1329                  0        0                   0             0   
1330                  0        0                   0             0   
1331                  0        0                   0             0   
1332                  0        0                   0             0   
1333                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1329               0                0                 0                    0   
1330               0                0                 0                    0   
1331               0                0                 0                    0   
1332               0                0                 0                    0   
1333               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  \
0                0         0  ...              0   
1                0         0  ...              0   
2                0         0  ...              0   
3                0         0  ...              0   
4                0         0  ...              0   
...            ...       ...  ...            ...   
1329             0         0  ...              0   
1330             0         0  ...              0   
1331             0         0  ...              0   
1332             0         0  ...              0   
1333             0         0  ...              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1329                              0                 0            0   
1330                              0                 0            0   
1331                              0                 0            0   
1332                              0                 0            0   
1333                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1329                      0                0         0            0   
1330                      0                0         0            0   
1331                      0                0         0            0   
1332                      0 

In [38]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  \
0                    0             0  ...              0   
1                    0             0  ...              0   
2                    0             0  ...              0   
3                    0             0  ...              0   
4                    0             0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 237 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [39]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0                 0.00           0.0  ...            0.0   
1               0.0                 0.00           0.0  ...            0.0   
2               0.0                 0.00           0.0  ...            0.0   
3               0.0                 0.00           0.0  ...            0.0   
4               0.0                 0.04           0.0  ...            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 237 columns]

Let's make a function to get the top most common venue categories

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [41]:
import numpy as np

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Breakfast Spot    College Auditorium   
1           Coffee Shop          Dance Studio        Sandwich Place   
2                  Bank                 Diner              Pharmacy   
3   Japanese Restaurant                  Bank                  Café   
4           Coffee Shop    Italian Restaurant        Breakfast Spot   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0             Drugstore            Donut Shop                    Dog Run   
1          Skating Rink                   Gym                        Pub   
2        Ice Cream Shop            Restaurant  Middle Eastern Restaurant   
3           Yoga Studio          Dance Studio                  Drugstore   
4       Thai Restaurant                   Pub                    Butcher   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Distribution Center        Discount Store                  Diner  
1      Department Store           Curling Ice          Deli / Bodega  
2     Mobile Phone Shop         Deli / Bodega         Sandwich Place  
3            Donut Shop               Dog Run    Distribution Center  
4                  Café      Sushi Restaurant   Fast Food Restaurant

Let's make the model to cluster our Neighbourhoods

In [43]:

# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [44]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 4, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1,
       1, 1, 1, 1, 1, 0], dtype=int32)

In [45]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [46]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels   1st Most Common Venue  \
0  43.753259 -79.329656             0.0       Food & Drink Shop   
1  43.725882 -79.315572             1.0             Coffee Shop   
2  43.654260 -79.360636             1.0             Coffee Shop   
3  43.718518 -79.464763             1.0  Furniture / Home Store   
4  43.662301 -79.389494             1.0             Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park           Yoga Studio            Donut Shop   
1  Portuguese Restaurant     French Restaurant          Hockey Arena   
2                   Park                Bakery        Breakfast Spot   
3         Clothing Store     Accessories Store    Athletics & Sports   
4                  Diner      Sushi Restaurant           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Dog Run   Distribution Center        Discount Store   
1           Pizza Place           College Gym              Creperie   
2               Theater                  Café            Restaurant   
3           Event Space    Miscellaneous Shop              Boutique   
4              Creperie        Sandwich Place         Smoothie Shop   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0                 Diner     Dim Sum Restaurant           Dessert Shop  
1   Distribution Center         Discount Store                  Diner  
2                   Pub         Chocolate Shop            Yoga Studio  
3           Coffee Shop  Vietnamese Restaurant       Airport Terminal  
4   Japanese Restaurant     Italian Restaurant           Burger Joint

Drop all the NaN values to prevent data skew

In [47]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [48]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [49]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0     Food & Drink Shop   
21              York             0.0                  Park   
35         East York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0                  Park   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park           Yoga Studio            Donut Shop   
21         Women's Store                  Pool      Cuban Restaurant   
35     Convenience Store           Yoga Studio          Cupcake Shop   
66     Convenience Store           Yoga Studio          Cupcake Shop   
85            Playground          Intersection           Yoga Studio   
91            Playground                 Trail              Creperie   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Dog Run   Distribution Center        Discount Store   
21            Donut Shop               Dog Run   Distribution Center   
35            Donut Shop               Dog Run   Distribution Center   
66            Donut Shop               Dog Run   Distribution Center   
85      Cuban Restaurant               Dog Run   Distribution Center   
91               Dog Run   Distribution Center        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Diner    Dim Sum Restaurant           Dessert Shop  
21        Discount Store                 Diner     Dim Sum Restaurant  
35        Discount Store                 Diner     Dim Sum Restaurant  
66        Discount Store                 Diner     Dim Sum Restaurant  
85        Discount Store                 Diner     Dim Sum Restaurant  
91                 Diner    Dim Sum Restaurant           Dessert Shop

Cluster 2

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
1          North York             1.0             Coffee Shop   
2    Downtown Toronto             1.0             Coffee Shop   
3          North York             1.0  Furniture / Home Store   
4    Downtown Toronto             1.0             Coffee Shop   
7          North York             1.0                     Gym   
..                ...             ...                     ...   
98          Etobicoke             1.0                   River   
99   Downtown Toronto             1.0             Coffee Shop   
100      East Toronto             1.0      Light Rail Station   
101         Etobicoke             1.0        Business Service   
102         Etobicoke             1.0             Social Club   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1    Portuguese Restaurant     French Restaurant          Hockey Arena   
2                     Park                Bakery        Breakfast Spot   
3           Clothing Store     Accessories Store    Athletics & Sports   
4                    Diner      Sushi Restaurant           Yoga Studio   
7              Coffee Shop   Sporting Goods Shop            Beer Store   
..                     ...                   ...                   ...   
98              Smoke Shop                  Pool           Yoga Studio   
99        Sushi Restaurant                  Park       Bubble Tea Shop   
100   Gym / Fitness Center  Fast Food Restaurant               Butcher   
101         Baseball Field     Electronics Store             Drugstore   
102      Convenience Store        Discount Store        Sandwich Place   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1             Pizza Place           College Gym              Creperie   
2                 Theater                  Café            Restaurant   
3             Event Space    Miscellaneous Shop              Boutique   
4                Creperie        Sandwich Place         Smoothie Shop   
7              Restaurant        Discount Store             Bike Shop   
..                    ...                   ...                   ...   
98    Distribution Center        Discount Store                 Diner   
99      Indian Restaurant              Beer Bar                 Diner   
100            Skate Park         Auto Workshop            Restaurant   
101            Donut Shop               Dog Run   Distribution Center   
102          Burger Joint  Fast Food Restaurant                Bakery   

    8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
1     Distribution Center         Discount Store                  Diner  
2                     Pub         Chocolate Shop            Yoga Studio  
3             Coffee Shop  Vietnamese Restaurant       Airport Terminal  
4     Japanese Restaurant     Italian Restaurant           Burger Joint  
7      Italian Restaurant             Smoke Shop         Sandwich Place  
..                    ...                    ...                    ...  
98     Dim Sum Restaurant           Dessert Shop       Department Store  
99              Bookstore              Juice Bar         Breakfast Spot  
100         Burrito Place       Recording Studio            Pizza Place  
101        Discount Store                  Diner     Dim Sum Restaurant  
102       Supplement Shop            Flower Shop         Hardware Store  

[89 rows x 12 columns]

Cluster 3

In [52]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             2.0                Bakery           Yoga Studio   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
11           Escape Room  Eastern European Restaurant             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11            Donut Shop               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
11        Discount Store                  Diner

Cluster 4

In [53]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough             3.0            Playground           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32      Cuban Restaurant            Donut Shop               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32   Distribution Center        Discount Store                 Diner   

   9th Most Common Venue 10th Most Common Venue  
32    Dim Sum Restaurant           Dessert Shop

Cluster 5

In [55]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             4.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6     Electronics Store             Drugstore            Donut Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6               Dog Run   Distribution Center        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
6                 Diner     Dim Sum Restaurant

We have successfully cluster Toronto neighbourhood based on venue categories!